### PR TITLE
fix: Tier module fixes and improvements (part 3)

### PR DESCRIPTION
**Pull request by iverc**
## Description

This PR aims to add a number of fixes and additional improvements on top of work done in https://github.com/sourcenetwork/sourcehub/pull/8

## Tasks
- [x] Improve `CancelUnlocking` logic to support partial lockups.
- [x] Improve `SubtractLockup` to remove existing records if subtracting whole amount.
- [x] Add `GetLockups` to fetch all lockups for a given `delAddr`.
- [x] Add `SubtractUnlockingLockup` to handle partial unlocking lockups.
- [x] Add `SaveLockup` to be used in  `SubtractUnlockingLockup` and `InitGenesis`.
- [x] Emit corresponding events in `Lock`, `Unlock`, `Redelegate`, and `CompleteUnlocking`.
- [x] Fix bugs in `TestTotalAmountByAddr`, `GetLockup`, and `removeUnlockingLockup`.
- [x] Add more keeper and lockup tests, fix and improve existing tests.
- [x] Add msg server test setup, `TestMsgUpdateParams`, `TestMsgLock`, `TestMsgUnock`, `TestMsgRedelegate`, and `TestMsgCancelUnlocking`.

### CODE REVIEW

File: x/tier/keeper/calculate_credit_test.go
Comment (jsimnz) [inline x/tier/keeper/calculate_credit_test.go:1]: *(commenting on this whole file as it reminded me of some other testing)*
todo: Great tests for `calculateCredit`, but noticed it wasn't handling the *within* epoch calculations. This is obviously done in another function `proratedCredit` but isn't well tested on its own, and this is one place where there was a bug (not fixed in a previous PR) in the old implementation.

Comment (jsimnz) [reply x/tier/keeper/calculate_credit_test.go:1]: Basically, something like this unit test structure, but including some sense of time w.r.t the epoch duration

Comment (jsimnz) [reply x/tier/keeper/calculate_credit_test.go:1]: There are notable tests related to this through the various keeper unit tests, but I think doing something a little more stateless and declarative vs procedural just to make sure the math is correct.

**Review by jsimnz — APPROVED** [review]
A quick change for some basic unit testing for the prorated credit calculation. 

Beyond that, this is a really great change. Appreciate the additional events, and love all the testing, really have a knack for that 👍

File: x/tier/keeper/calculate_credit_test.go
Comment (iverc) [reply x/tier/keeper/calculate_credit_test.go:1]: Thank you! I'll include all the tests for `proratedCredit`, `burnAllCredits`, and `resetAllCredits` in the part 4 PR.

**Pull request merged**